util/syspolicy/internal/metrics/metrics.go

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package metrics provides logging and reporting for policy settings and scopes.
+package metrics
+
+import (
+	"strings"
+	"sync"
+
+	xmaps "golang.org/x/exp/maps"
+
+	"tailscale.com/syncs"
+	"tailscale.com/types/lazy"
+	"tailscale.com/util/clientmetric"
+	"tailscale.com/util/mak"
+	"tailscale.com/util/syspolicy/internal"
+	"tailscale.com/util/syspolicy/internal/loggerx"
+	"tailscale.com/util/syspolicy/setting"
+	"tailscale.com/util/testenv"
+)
+
+var lazyReportMetrics lazy.SyncValue[bool] // used as a test hook
+
+// ShouldReport reports whether metrics should be reported on the current environment.
+func ShouldReport() bool {
+	return lazyReportMetrics.Get(func() bool {
+		// macOS, iOS and tvOS create their own metrics,
+		// and we don't have syspolicy on any other platforms.
+		return setting.PlatformList{"android", "windows"}.HasCurrent()
+	})
+}
+
+// Reset metrics for the specified policy origin.
+func Reset(origin *setting.Origin) {
+	scopeMetrics(origin).Reset()
+}
+
+// ReportConfigured updates metrics and logs that the specified setting is
+// configured with the given value in the origin.
+func ReportConfigured(origin *setting.Origin, setting *setting.Definition, value any) {
+	settingMetricsFor(setting).ReportValue(origin, value)
+}
+
+// ReportError updates metrics and logs that the specified setting has an error
+// in the origin.
+func ReportError(origin *setting.Origin, setting *setting.Definition, err error) {
+	settingMetricsFor(setting).ReportError(origin, err)
+}
+
+// ReportNotConfigured updates metrics and logs that the specified setting is
+// not configured in the origin.
+func ReportNotConfigured(origin *setting.Origin, setting *setting.Definition) {
+	settingMetricsFor(setting).Reset(origin)
+}
+
+// metric is an interface implemented by [clientmetric.Metric] and [funcMetric].
+type metric interface {
+	Add(v int64)
+	Set(v int64)
+}
+
+// policyScopeMetrics are metrics that apply to an entire policy scope rather
+// than a specific policy setting.
+type policyScopeMetrics struct {
+	hasAny     metric
+	numErrored metric
+}
+
+func newScopeMetrics(scope setting.Scope) *policyScopeMetrics {
+	prefix := metricScopeName(scope)
+	if prefix != "" {
+		prefix += "_"
+	}
+	// {os}_syspolicy_{scope_unless_device}_any
+	// Example: windows_syspolicy_any or windows_syspolicy_user_any.
+	hasAny := newMetric(prefix+"any", clientmetric.TypeGauge)
+	// {os}_syspolicy_{scope_unless_device}_errors
+	// Example: windows_syspolicy_errors or windows_syspolicy_user_errors.
+	//
+	// TODO(nickkhyl): maybe make the `{os}_syspolicy_errors` metric a gauge rather than a counter?
```

Comment:
Using a counter, we'll report the number of times a resultant policy is changed multiplied by the total number of errored policy settings in the applicable policy stores.

For example, if there are two policy settings with invalid values and we re-read the policy 10 times, we'll report 20 new errors. The number of times we've re-read the policy within an hour, day, or week is out of our control and is heavily influenced by the environment and user behavior. Therefore, I'm struggling to understand why a counter is more appropriate for the `windows_syspolicy_errors` metric, especially since all other `syspolicy` metrics are gauges.

Reply:
Yeah, now I'm unsure of the value in having this specific metric at all 😁 

I think my overall feeling is that the "number of policies that failed to load" is less useless than the "number of times an error occurred", even without knowing the denominator for the first one. The second metric could be greatly skewed by a customer with ephemeral nodes or a customer who has an unusually short Group Policy processing interval (e.g., 1 minute vs. the standard 90 minutes), and so on.

I think I'll keep it as is for now with the TODO in place, but add a link to this thread from the code.

Also, maybe @kabirsikand has any insights on how the `windows_syspolicy_errors` and `macos_syspolicy_errors`/`ios_syspolicy_errors` metrics are used?